cmd/update-main.go

```diff
@@ -68,90 +68,108 @@ const (
 	minioReleaseURL           = "https://dl.minio.io/server/minio/release/" + runtime.GOOS + "-" + runtime.GOARCH + "/"
 )
 
-func getCurrentReleaseTime(minioVersion, minioBinaryPath string) (releaseTime time.Time, err error) {
-	if releaseTime, err = time.Parse(time.RFC3339, minioVersion); err == nil {
+// updateCheckParams - collection of parameters used to check if the
+// version of the current server process is older than the latest
+// available official release. They enable the testability of various
+// routines in the update checking process.
+type updateCheckParams struct {
+	// version of the current server process
+	version string
+	// path to the server binary
+	serverBinaryPath string
+	// file on the server to check for docker env
+	cgroupFile string
+	// URL of release checksum file (contains version of latest
+	// release)
+	releaseChecksumURL string
+	// current process run mode
+	serverMode string
+}
+
+// getCurrentReleaseTime - returns this process's release time.  If it
+// is official minio version, parsed version is returned else minio
+// binary's mod time is returned.
+func (u *updateCheckParams) getCurrentReleaseTime() (releaseTime time.Time,
```

Comment:
From [here](https://github.com/golang/go/wiki/CodeReviewComments#receiver-type):

> If the receiver is a large struct or array, a pointer receiver is more efficient. How large is large? Assume it's equivalent to passing all its elements as arguments to the method. If that feels too large, it's also too large for the receiver.

We'll need some sort of convention to decide when a struct is considered "large". We can check with AB on this.

> If the receiver is a small array or struct that is naturally a value type (for instance, something like the time.Time type), with no mutable fields and no pointers, or is just a simple basic type such as int or string, a value receiver makes sense.

This suggests that for anything other than basic types like int, string or a "naturally value type" like time.Time we should use pointer receiver.

> A value receiver can reduce the amount of garbage that can be generated; if a value is passed to a value method, an on-stack copy can be used instead of allocating on the heap. (The compiler tries to be smart about avoiding this allocation, but it can't always succeed.) Don't choose a value receiver type for this reason without profiling first.


_Don't choose a value receiver type for this reason without profiling first._

Very confusing, are they saying use pointers even for basic types like int, string and then perform profiling and then change it to value receivers if there is a performance problem?